File: PhotonCore/src/main/java/com/outoftheboxrobotics/photoncore/ReflectionUtils.java
package com.outoftheboxrobotics.photoncore;

import java.lang.reflect.Field;

@SuppressWarnings("rawtypes")
public class ReflectionUtils {
    public static Field getField(Class clazz, String fieldName) {
        try {
            Field field = clazz.getDeclaredField(fieldName);
            field.setAccessible(true);
            return field;
        } catch (NoSuchFieldException e) {
            Class superClass = clazz.getSuperclass();
            if (superClass != null) {
                return getField(superClass, fieldName);
            }
        }
        return null;
    }

    @SuppressWarnings({"unchecked"})
    public static <T> T getFieldValue(Object object, String fieldName)  {
        Field field = getField(object.getClass(), fieldName);
        if (field != null) {
            field.setAccessible(true);
            try{
                return (T)field.get(object);
            }catch (Exception e)
            {
                return null;
            }

        }
        return null;
    }
    public static void setFieldValue(Object object, Object value, String fieldName)  {
        Field field = getField(object.getClass(), fieldName);
        if (field != null) {
            field.setAccessible(true);
            try{
                field.set(object, value);
            }catch (Exception ignored)
            {

            }

        }
    }
    public static void deepCopy(Object org, Object target){
        Field[] fields = org.getClass().getDeclaredFields();
        for(Field f : fields){
            f.setAccessible(true);
            Field f2 = getField(target.getClass(), f.getName());
            if(f2 != null){
                f2.setAccessible(true);
                try {
                    f2.set(target, f.get(org));
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: PhotonCore/src/main/java/com/outoftheboxrobotics/photoncore/ReflectionUtils.kt
package com.outoftheboxrobotics.photoncore

import java.lang.reflect.Field

object ReflectionUtils {
    @JvmStatic
    fun getField(clazz: Class<*>, fieldName: String): Field? {
        try {
            val f = clazz.getDeclaredField(fieldName)
            f.isAccessible = true
            return f
        } catch (e: NoSuchFieldException) {
            val superClass = clazz.superclass
            if (superClass != null) {
                return getField(superClass, fieldName)
            }
        }
        return null
    }

    fun getField(clazz: Class<*>, target: Class<*>): Field? {
        for (f in clazz.declaredFields) {
            if (f.type == target) {
                f.isAccessible = true
                return f
            }
        }
        val superClass = clazz.superclass
        return if (superClass != null) {
            getField(clazz.superclass, target)
        } else {
            null
        }
    }

    fun deepCopy(org: Any, target: Any) {
        val fields = org.javaClass.declaredFields
        for (f in fields) {
            f.isAccessible = true
            val f2 = getField(target.javaClass, f.name)
            if (f2 != null) {
                f2.isAccessible = true
                try {
                    f2[target] = f[org]
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                }
            }
        }
    }
}